torchsde/_brownian/derived.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import brownian_base
+from . import brownian_interval
+from ..types import Optional, Scalar, Tensor
+
+
+class ReverseBrownian(brownian_base.BaseBrownian):
+    def __init__(self, base_brownian):
+        super(ReverseBrownian, self).__init__()
+        self.base_brownian = base_brownian
+
+    def __call__(self, ta, tb=None, return_U=False, return_A=False):
+        # Whether or not to negate the statistics depends on the return value of the adjoint SDE. Currently, the adjoint
+        # returns negated drift and diffusion, so we don't negate here.
+        return self.base_brownian(-tb, -ta, return_U=return_U, return_A=return_A)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(base_brownian={self.base_brownian})"
+
+    @property
+    def dtype(self):
+        return self.base_brownian.dtype
+
+    @property
+    def device(self):
+        return self.base_brownian.device
+
+    @property
+    def shape(self):
+        return self.base_brownian.shape
+
+    @property
+    def levy_area_approximation(self):
+        return self.base_brownian.levy_area_approximation
+
+
+class BrownianPath(brownian_interval.BrownianInterval):
+    """Brownian path, storing every computed value.
+
+    Useful for speed, when memory isn't a concern.
+
+    To use:
+    >>> bm = BrownianPath(t0=0.0, w0=torch.zeros(4, 1))
+    >>> bm(0., 0.5)
+    tensor([[ 0.0733],
+            [-0.5692],
+            [ 0.1872],
+            [-0.3889]])
+    """
+
+    def __init__(self, t0: Scalar, w0: Tensor, window_size: int = 8):
```

Comment:
Could we also try support the new API at the same time? Something like `__init__(self, t0, t1, shape..., w0=None, window_size=...)`? The legacy `w0` and `window_size` should mostly be there to not anger people who might have code depending on this codebase.

Reply:
The problem then is matching up positional arguments. It ends up having to be something like `__init__(self, t0, w0, window_size, ...)`, with some complicated set of checks to try and figure out if they're using old or new style API - in the latter case then re-interpreting `w0` as `t1`, and so on.

Moreover there's no real need to do this - if someone wants to use the new-style API then they can just use BInterval directly.